hikari/internal/messages.py

```diff
@@ -0,0 +1,161 @@
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+"""Utility functions used for managing messages on Discord."""
+from __future__ import annotations
+
+__all__: typing.Sequence[str] = ("build_message_payload",)
+
+import typing
+
+from hikari import embeds as embeds_
+from hikari import files
+from hikari import undefined
+from hikari.internal import data_binding
+
+if typing.TYPE_CHECKING:
+    from hikari import guilds
+    from hikari import messages as messages_
+    from hikari import snowflakes
+    from hikari import users
+    from hikari.api import entity_factory as entity_factory_
+    from hikari.api import special_endpoints
+
+    class _AttachmentPayload(typing.TypedDict):
+        filename: str
+        id: int
+
+
+def _extend_attachments(
+    payloads: typing.List[_AttachmentPayload],
+    to_upload: typing.Dict[int, files.Resource[typing.Any]],
+    edit: bool,
+    attachments: typing.Sequence[typing.Union[files.Resourceish, messages_.Attachment]],
+) -> None:
+    attachment_id = 0
+    for attachment in attachments:
+        if edit and isinstance(attachment, messages_.Attachment):
+            payloads.append({"id": attachment.id, "filename": attachment.filename})
+
+        else:
+            resource = files.ensure_resource(attachment)
+            to_upload[attachment_id] = resource
+            payloads.append({"id": attachment_id, "filename": resource.filename})
+            attachment_id += 1
+
+
+def build_message_payload(
+    entity_factory: entity_factory_.EntityFactory,
+    /,
+    *,
+    content: undefined.UndefinedOr[typing.Any] = undefined.UNDEFINED,
+    attachments: undefined.UndefinedNoneOr[
+        typing.Sequence[typing.Union[files.Resourceish, messages_.Attachment]]
+    ] = undefined.UNDEFINED,
+    components: undefined.UndefinedNoneOr[typing.Sequence[special_endpoints.ComponentBuilder]] = undefined.UNDEFINED,
+    embeds: undefined.UndefinedNoneOr[typing.Sequence[embeds_.Embed]] = undefined.UNDEFINED,
+    flags: typing.Union[undefined.UndefinedType, int, messages_.MessageFlag] = undefined.UNDEFINED,
+    tts: undefined.UndefinedOr[bool] = undefined.UNDEFINED,
+    mentions_everyone: undefined.UndefinedOr[bool] = undefined.UNDEFINED,
+    mentions_reply: undefined.UndefinedOr[bool] = undefined.UNDEFINED,
+    user_mentions: undefined.UndefinedOr[
+        typing.Union[snowflakes.SnowflakeishSequence[users.PartialUser], bool]
+    ] = undefined.UNDEFINED,
+    role_mentions: undefined.UndefinedOr[
+        typing.Union[snowflakes.SnowflakeishSequence[guilds.PartialRole], bool]
+    ] = undefined.UNDEFINED,
+    edit: bool = False,
+) -> typing.Tuple[data_binding.JSONObjectBuilder, typing.Mapping[int, files.Resource[typing.Any]]]:
+    """Create a message create/edit payload."""
+    to_upload: typing.Dict[int, files.Resource[typing.Any]] = {}
```

Comment:
Could've done `typing.List[typing.Tuple[int, files.Resource[typing.Any]]` but stuck with Dict/Mapping for now since it keeps the type cleaner and this does reach the public interface a bit